File: app/menu.js
const { app, dialog, nativeTheme, Menu } = require('electron');
const fs = require('fs');
const path = require('path');
const { saveLastOpenedFile, getRecentFiles } = require('./utils');
const { DEBUG } = require('./config');

function buildRecentFilesSubmenu(mainWindow) {
    const recentFiles = getRecentFiles();
    DEBUG.log('Building recent files submenu with files:', recentFiles);
    
    if (recentFiles.length === 0) {
        return [{
            label: 'No Recent Files',
            enabled: false
        }];
    }
    
    const submenu = recentFiles.map(filePath => ({
        label: path.basename(filePath),
        click: () => {
            try {
                const content = fs.readFileSync(filePath, 'utf-8');
                saveLastOpenedFile(filePath);
                mainWindow.webContents.send('file-opened', {
                    filePath,
                    content
                });
                // Rebuild the entire menu after opening a recent file
                const template = createMenu(mainWindow);
                const menu = Menu.buildFromTemplate(template);
                Menu.setApplicationMenu(menu);
            } catch (error) {
                DEBUG.error('Error opening recent file:', error);
                dialog.showErrorBox('Error', `Could not open file: ${filePath}`);
            }
        }
    }));
    
    submenu.push(
        { type: 'separator' },
        {
            label: 'Clear Recent Files',
            click: () => {
                saveLastOpenedFile(null);
                // Rebuild the entire menu after clearing
                const template = createMenu(mainWindow);
                const menu = Menu.buildFromTemplate(template);
                Menu.setApplicationMenu(menu);
            }
        }
    );
    
    return submenu;
}

function createMenu(mainWindow) {
    const isMac = process.platform === "darwin";

    const template = [
        ...(isMac
            ? [{
                label: app.name,
                submenu: [
                    { role: "about" },
                    { type: "separator" },
                    { role: "services" },
                    { type: "separator" },
                    { role: "hide" },
                    { role: "hideOthers" },
                    { role: "unhide" },
                    { type: "separator" },
                    { role: "quit" },
                ],
            }]
            : []),
        {
            label: "File",
            submenu: [
                {
                    label: "New",
                    accelerator: isMac ? "Cmd+N" : "Ctrl+N",
                    click: async () => {
                        try {
                            const { filePath } = await dialog.showSaveDialog(mainWindow, {
                                title: "Create New Markdown File",
                                defaultPath: "untitled.md",
                                filters: [{ name: "Markdown Files", extensions: ["md"] }],
                            });

                            if (filePath) {
                                fs.writeFileSync(filePath, ""); // Create an empty markdown file
                                saveLastOpenedFile(filePath);
                                mainWindow.webContents.send("file-opened", {
                                    filePath,
                                    content: "",
                                });
                                // Rebuild menu after creating new file
                                const template = createMenu(mainWindow);
                                const menu = Menu.buildFromTemplate(template);
                                Menu.setApplicationMenu(menu);
                            }
                        } catch (error) {
                            DEBUG.error("Error creating new file:", error);
                        }
                    },
                },
                {
                    label: "Open",
                    accelerator: isMac ? "Cmd+O" : "Ctrl+O",
                    click: async () => {
                        try {
                            const { canceled, filePaths } = await dialog.showOpenDialog(
                                mainWindow,
                                {
                                    properties: ["openFile"],
                                    filters: [{ name: "Markdown Files", extensions: ["md"] }],
                                }
                            );
                            if (!canceled && filePaths.length > 0) {
                                const filePath = filePaths[0];
                                const content = fs.readFileSync(filePath, "utf-8");
                                saveLastOpenedFile(filePath);
                                mainWindow.webContents.send("file-opened", {
                                    filePath,
                                    content,
                                });
                                // Rebuild menu after opening file
                                const template = createMenu(mainWindow);
                                const menu = Menu.buildFromTemplate(template);
                                Menu.setApplicationMenu(menu);
                            }
                        } catch (error) {
                            DEBUG.error("Error opening file:", error);
                        }
                    },
                },
                {
                    label: "Recent Files",
                    submenu: buildRecentFilesSubmenu(mainWindow)
                },
                { type: "separator" },
                {
                    label: "Save",
                    accelerator: isMac ? "Cmd+S" : "Ctrl+S",
                    click: () => {
                        mainWindow.webContents.send("file-save-request");
                    },
                },
                isMac ? { role: "close" } : { role: "quit" },
            ],
        },
        {
            label: "Edit",
            submenu: [
                { role: "undo" },
                { role: "redo" },
                { type: "separator" },
                { role: "cut" },
                { role: "copy" },
                { role: "paste" },
                { role: "selectAll" },
            ],
        },
        {
            label: "View",
            submenu: [
                {
                    label: "Dark Mode",
                    type: "radio",
                    click: () => {
                        nativeTheme.themeSource = "dark";
                        mainWindow.webContents.send("update-dark-mode", true);
                    },
                },
                {
                    label: "Light Mode",
                    type: "radio",
                    click: () => {
                        nativeTheme.themeSource = "light";
                        mainWindow.webContents.send("update-dark-mode", false);
                    },
                },
                {
                    label: "System Theme",
                    type: "radio",
                    click: () => {
                        nativeTheme.themeSource = "system";
                        mainWindow.webContents.send(
                            "update-dark-mode",
                            nativeTheme.shouldUseDarkColors
                        );
                    },
                },
                { type: "separator" },
                { role: "reload" },
                { role: "toggledevtools" },
            ],
        },
        {
            label: "Window",
            submenu: [{ role: "minimize" }, { role: "zoom" }, { role: "close" }],
        },
        {
            label: "Help",
            submenu: [
                {
                    label: "About Lexorium",
                    click: () => {
                        dialog.showMessageBox({
                            type: 'none',
                            title: 'About Lexorium',
                            message: `Lexorium \n\nVersion: ${app.getVersion()}\n\nVisit our GitHub page:`,
                            buttons: ['Open GitHub', 'Close'],
                            defaultId: 0,
                            icon: 'assets/icon.png',
                        }).then(result => {
                            if (result.response === 0) {
                                const { shell } = require("electron");
                                shell.openExternal("https://github.com/antnsn/lexorium");
                            }
                        });
                    },
                },
            ],
        },
    ];

    return template;
}

// Update the menu when a file is saved
function updateMenu(mainWindow) {
    const template = createMenu(mainWindow);
    const menu = Menu.buildFromTemplate(template);
    Menu.setApplicationMenu(menu);
}

module.exports = { createMenu, updateMenu };

File: app/src-tauri/src/menu.rs
use tauri::{
    menu::{MenuBuilder, MenuItemBuilder, SubmenuBuilder, CheckMenuItemBuilder},
    App, AppHandle, Emitter,
};

pub fn create_menu(app: &App) -> Result<(), Box<dyn std::error::Error>> {
    let handle = app.handle();

    // ── App submenu (macOS "About" slot) ──
    let app_menu = SubmenuBuilder::new(handle, "Lexorium")
        .about(None)
        .separator()
        .item(&MenuItemBuilder::with_id("settings", "Settings…")
            .accelerator("CmdOrCtrl+,")
            .build(handle)?)
        .separator()
        .services()
        .separator()
        .hide()
        .hide_others()
        .show_all()
        .separator()
        .quit()
        .build()?;

    // ── File ──
    let file_menu = SubmenuBuilder::new(handle, "File")
        .item(&MenuItemBuilder::with_id("file-new", "New Document")
            .accelerator("CmdOrCtrl+N")
            .build(handle)?)
        .item(&MenuItemBuilder::with_id("file-open", "Open…")
            .accelerator("CmdOrCtrl+O")
            .build(handle)?)
        .separator()
        .item(&MenuItemBuilder::with_id("file-save", "Save")
            .accelerator("CmdOrCtrl+S")
            .build(handle)?)
        .item(&MenuItemBuilder::with_id("file-save-as", "Save As…")
            .accelerator("CmdOrCtrl+Shift+S")
            .build(handle)?)
        .separator()
        .close_window()
        .build()?;

    // ── Edit ──
    let edit_menu = SubmenuBuilder::new(handle, "Edit")
        .undo()
        .redo()
        .separator()
        .cut()
        .copy()
        .paste()
        .select_all()
        .build()?;

    // ── View ──
    let dark_mode_item = CheckMenuItemBuilder::new("Dark Mode")
        .id("toggle-dark-mode")
        .accelerator("CmdOrCtrl+Shift+D")
        .build(handle)?;

    let view_menu = SubmenuBuilder::new(handle, "View")
        .item(&dark_mode_item)
        .separator()
        .item(&MenuItemBuilder::with_id("toggle-sidebar", "Toggle Sidebar")
            .accelerator("CmdOrCtrl+\\")
            .build(handle)?)
        .build()?;

    // ── Help ──
    let help_menu = SubmenuBuilder::new(handle, "Help")
        .item(&MenuItemBuilder::with_id("open-repo", "Lexorium on GitHub")
            .build(handle)?)
        .build()?;

    let menu = MenuBuilder::new(handle)
        .item(&app_menu)
        .item(&file_menu)
        .item(&edit_menu)
        .item(&view_menu)
        .item(&help_menu)
        .build()?;

    app.set_menu(menu)?;

    // ── Handle menu events ──
    app.on_menu_event(move |app_handle, event| {
        let id = event.id().as_ref();
        log::info!("Menu event fired: {}", id);
        match id {
            "file-new" => {
                log::info!("Emitting file-new event");
                let _ = app_handle.emit("file-new", "new");
            }
            "file-open" => {
                log::info!("Emitting file-open-request event");
                let _ = app_handle.emit("file-open-request", "open");
            }
            "file-save" => {
                log::info!("Emitting file-save-request event");
                let _ = app_handle.emit("file-save-request", "save");
            }
            "file-save-as" => {
                log::info!("Emitting file-save-as-request event");
                let _ = app_handle.emit("file-save-as-request", "save-as");
            }
            "settings" => {
                log::info!("Emitting show-settings event");
                let _ = app_handle.emit("show-settings", "settings");
            }
            "toggle-dark-mode" => {
                log::info!("Emitting dark-mode-toggle event");
                let _ = app_handle.emit("dark-mode-toggle", "toggle");
            }
            "toggle-sidebar" => {
                log::info!("Emitting toggle-sidebar event");
                let _ = app_handle.emit("toggle-sidebar", "sidebar");
            }
            "open-repo" => {
                open_url(app_handle, "https://github.com/antnsn/Lexorium");
            }
            _ => {
                log::debug!("Unhandled menu event: {}", id);
            }
        }
    });

    Ok(())
}

fn open_url(_app: &AppHandle, url: &str) {
    let _ = open::that(url);
}
